docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,136 @@
+
+This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
+From a preliminary investigation, the risk is that the manager that created the object and the new SSA manager will become co-owner of the same labels, and this will prevent deletion of labels.
+
+The synchronization process will be implemented in the Machine controller. Reconciliation is triggered both when Machine object changes, or when the Node changes (the machine controller watches Nodes within the workload cluster).
+
+#### Delay between Node Create and Label Sync
+
+The Node object is first created when kubelet joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the machine controller kicks in to apply the labels on the Node.
+
+Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.
+
+In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.
+
+One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `node.cluster.x-k8s.io=uninitialized:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).
+
+The process of tainting the nodes, can be carried out by the user and can be documented as follows:
+
+```
+If you utilize inequality based selection for workload placement, to prevent unintended scheduling of pods during the initial node startup phase, it is recommend that you specify the following taint in your KubeadmConfigTemplate:
+`node.cluster.x-k8s.io=uninitialized:NoSchedule`
+```
+
+After the node has come up and the machine controller has applied the labels, the machine controller will also remove this specific taint if it's present.
+
+During the implementation we will consider also automating the insertion of the taint via CABPK in order to simplify UX.
```

Comment:
This proposal will also affect 3rd party bootstrap providers, should we make taint insertion a provider contract in case it will be automated in CAPBK?

Reply:
yes, if we go down this path this should be documented in the contract as optional requirement. I will add a note

Reply:
done!